substrate/primitives/runtime/src/type_with_default.rs

```diff
@@ -0,0 +1,506 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides a type that wraps another type and provides a default value.
+
+use crate::traits::{Bounded, One, Zero};
+use codec::{Compact, CompactAs, Decode, Encode, HasCompact, MaxEncodedLen};
+use core::{
+	fmt::Display,
+	marker::PhantomData,
+	ops::{
+		Add, AddAssign, BitAnd, BitOr, BitXor, Deref, Div, DivAssign, Mul, MulAssign, Not, Rem,
+		RemAssign, Shl, Shr, Sub, SubAssign,
+	},
+};
+use num_traits::{
+	CheckedAdd, CheckedDiv, CheckedMul, CheckedNeg, CheckedRem, CheckedShl, CheckedShr, CheckedSub,
+	Num, NumCast, PrimInt, Saturating, ToPrimitive,
+};
+use scale_info::TypeInfo;
+use sp_core::Get;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// A type that wraps another type and provides a default value.
+///
+/// Passes through arithmetical and many other operations to the inner value.
+#[derive(Encode, Decode, TypeInfo, Debug, MaxEncodedLen)]
```

Comment:
Crucially, this type should have not derived `TypeInfo`. It should have had a manual `TypeInfo` implementation that would mimic the internal `T`.